File: out_go_v5/git.go
package main

import (
	"fmt"
	"io/ioutil"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/plumbing"
	"github.com/go-git/go-git/v5/plumbing/storer"
	"github.com/go-git/go-git/v5/storage/memory"
)

type GitRepo struct {
	rootPath string
	repo     *git.Repository
}

type CommitSummary struct {
	id        string
	summary   string
	author    string
	timestamp int64
}

type CommitInfo struct {
	id        string
	summary   string
	author    string
	date      string
}

func NewGitRepo(rootPath string) (*GitRepo, error) {
	repo, err := git.Clone(memory.NewReturnEmptyStorage(), filepath.Dir(rootPath), &git.CloneOptions{
		URL:                   "",
		Depth:                 1,
		SingleBranch:          true,
		RecurseSubmodules:     git.DefaultRecursionDegree,
		RemoteName:            "origin",
		Bare:                  false,
		Progress:              true,
		DepthSpec:             "",
		NoHardwireCredentials: true,
	})
	if err != nil {
		return nil, err
	}
	workdir := filepath.Dir(rootPath)
	err = git.Clean(repo, workdir)
	if err != nil {
		return nil, err
	}
	return &GitRepo{rootPath: filepath.Dir(rootPath), repo: repo}, nil
}

func (g *GitRepo) RecentCommitCount(limit int) int {
	walk := g.repo.RevWalk(storer.NewTreeWalk(g.repo.Storage()))
	if walk.Step(&plumbing.NewCommit(walk)) {
		if !walk.Step(&plumbing.NewCommit(walk)) {
			return 0
		}
	}
	return walk.Take(limit).Count()
}

func (g *GitRepo) TotalCommitCount() int {
	walk := g.repo.RevWalk(storer.NewTreeWalk(g.repo.Storage()))
	if walk.Step(&plumbing.NewCommit(walk)) {
		if !walk.Step(&plumbing.NewCommit(walk)) {
			return 0
		}
	}
	return walk.Count()
}

func (g *GitRepo) UncommittedFileCount() int {
	statuses, err := g.repo.Status("", &git.StatusOptions{})
	if err != nil {
		return 0
	}
	return len(statuses)
}

func (g *GitRepo) ChangedFiles() []string {
	statuses, err := g.repo.Status("")
	if err != nil {
		return []string{}
	}
	files := []string{}
	for _, entry := range statuses {
		if entry.IsWorktreeModified() {
			path := filepath.Clean(entry.Filespace())
			files = append(files, path)
		}
	}
	return files
}

func (g *GitRepo) RootPath() string {
	return g.rootPath
}

func (g *GitRepo) GetHeadCommit() (string, error) {
	head, err := g.repo.Head()
	if err != nil {
		return "", err
	}
	headOid, err := head.Name()
	if err != nil {
		return "", err
	}
	cmd := exec.Command("git", "rev-parse", headOid)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return "", err
	}
	return string(out), nil
}

func (g *GitRepo) GetDiffStats(fromCommit string) (int, int, error) {
	cmd := exec.Command("git", "diff", "--numstat", fromCommit)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return 0, 0, err
	}
	output := string(out)
	addedLine := 0
	deletedLine := 0
	for _, line := range strings.Split(output, "\n") {
		parts := strings.Fields(line)
		if len(parts) >= 2 {
			if i, err := strconv.Atoi(parts[0]); err == nil {
				addedLine += i
			}
			if i, err := strconv.Atoi(parts[1]); err == nil {
				deletedLine += i
			}
		}
	}
	return addedLine, deletedLine, nil
}

func (g *GitRepo) GetCommitsByAuthor(authorName string, limit int) (string, error) {
	output, err := g.run("git", "log", "--author="+authorName, "-p", fmt.Sprintf("-n%d", limit))
	if err != nil {
		return "", err
	}
	return output, nil
}

func (g *GitRepo) LastCommitInfo() (CommitInfo, error) {
	output, err := g.run("git", "log", "-1", "--format=%H|%an|%ad|%s", "--date=iso-strict")
	if err != nil {
		return CommitInfo{}, err
	}
	commitInfo := strings.Split(string(output), "|")
	if len(commitInfo) != 4 {
		return CommitInfo{}, fmt.Errorf("Invalid commit information")
	}
	return CommitInfo{
		id:        commitInfo[0],
		summary:   commitInfo[3],
		author:    commitInfo[1],
		date:      commitInfo[2],
	}, nil
}

func (g *GitRepo) RecentCommits(limit int) ([]CommitSummary, error) {
	commitSummaries := []CommitSummary{}
	tree := storer.NewTreeWalk(g.repo.Storage())
	commitOid, err := g.repo.Head().Name()
	if err != nil {
		return nil, err
	}
	if err := tree.Step(&plumbing.NewCommit(g.repo.Storage(), commitOid)); err != nil {
		return nil, err
	}
	for limit > 0 {
		commitOid, err := tree.Parent(0)
		if err != nil {
			break
		}
		commit, err := g.repo.FindCommit(commitOid)
		if err != nil {
			break
		}
		commitSummaries = append(commitSummaries, g.SummarizeCommit(commit))
		limit--
	}
	return commitSummaries, nil
}

func (g *GitRepo) RecentCommitsForPath(relPath string, limit int) ([]CommitSummary, error) {
	tree := storer.NewTreeWalk(g.repo.Storage())
	commitOid, err := g.repo.Head().Name()
	if err != nil {
		return nil, err
	}
	if err := tree.Step(&plumbing.NewCommit(g.repo.Storage(), commitOid)); err != nil {
		return nil, err
	}
	commitSummaries := []CommitSummary{}
	for limit > 0 {
		commitOid, err := tree.Parent(0)
		if err != nil {
			break
		}
		commit, err := g.repo.FindCommit(commitOid)
		if err != nil {
			break
		}
		if g.CommitTouchesPath(commit, relPath) {
			commitSummaries = append(commitSummaries, g.SummarizeCommit(commit))
			limit--
		}
	}
	return commitSummaries, nil
}

func (g *GitRepo) CommitTouchesPath(commit *git.Commit, relPath string) bool {
	tree := commit.Tree()
	parent := commit.Parent(0)
	parentTree := parent.Tree()
	diff := g.repo.DiffTreeToTree(parentTree, tree, &git.DiffOptions{
		IncludeUntracked: true,
	})
	diff.ForEach(
		g.commitDiffHandler(g.rootPath, relPath, diff),
		&storer.TreeWalkConfig{Tree: parentTree},
		&storer.TreeWalkConfig{Tree: tree},
		&storer.TreeWalkConfig{Tree: tree},
	)
	return diff.HasTouched
}

func (g *GitRepo) SummarizeCommit(commit *git.Commit) CommitSummary {
	id := shortHash(commit.ID())
	summary := commit.Summary()
	author := commit.Author().Name()
	timestamp := commit.Time().Unix()

	return CommitSummary{
		id:        id,
		summary:   summary,
		author:    author,
		timestamp: timestamp,
	}
}

func (g *GitRepo) run(cmd string, args ...string) ([]byte, error) {
	parts := append([]string{cmd}, args...)
	output, err := exec.Command(parts[0], parts[1:]...).CombinedOutput()
	if err != nil {
		return nil, err
	}
	return output, nil
}

func shortHash(hash *plumbing.Hash) string {
	str := hash.String()
	return str[:8]
}

func (g *GitRepo) commitDiffHandler(rootPath string, relPath string, diff *git.Diff) func(delta *git.DiffDelta, out bool) bool {
	return func(delta *git.DiffDelta, out bool) bool {
		if newFile := delta.NewFile(); newFile != nil {
			if newFile.Path().String() != filepath.Join(rootPath, relPath) {
				return true
			}
		}
		if oldFile := delta.OldFile(); oldFile != nil {
			if oldFile.Path().String() != filepath.Join(rootPath, relPath) {
				return true
			}
		}
		return false
	}
}

func main() {
	g := &GitRepo{}
}

File: codex_ts_core3/src/git.ts
// TODO: use std::path::{Path, PathBuf};
// TODO: use std::process::Command;

// TODO: use anyhow::{anyhow, Result};
// TODO: use git2::{Commit, Oid, Repository};

export interface GitRepo {
  root: any;
  repo: any;
}


export interface CommitSummary {
  id: string;
  summary: string;
  author: string;
  time: number;
}


// impl GitRepo {
export function discover(root: any): any {
  // let repo = Repository::discover(root)?;
  // let workdir = repo
  // .workdir()
  // .ok_or_else(|| anyhow!("repository has no workdir"))?;
  // Ok(Self {
  // root: workdir.to_path_buf(),
  // repo,
  // })
}


export function recent_commit_count(&self: any, limit: number): number {
  // let mut revwalk = match self.repo.revwalk() {
  // Ok(rw) => rw,
  // Err(_) => return 0,
  // };
  // if revwalk.push_head().is_err() {
  // return 0;
  // }
  // revwalk.take(limit).count()
}


export function uncommitted_file_count(&self: any): number {
  // let statuses = match self.repo.statuses(None) {
  // Ok(s) => s,
  // Err(_) => return 0,
  // };
  // statuses.len()
}


export function root_path(&self: any): any {
  // &self.root
}


// /// Returns the current HEAD commit hash.
export function get_head_commit(&self: any): any {
  // let output = Command::new("git")
  // .current_dir(&self.root)
  // .args(&["rev-parse", "HEAD"])
  // .output()?;
  // 
  // if output.status.success() {
  // let hash = String::from_utf8_lossy(&output.stdout).trim().to_string();
  // Ok(hash)
  // } else {
  // Err(anyhow::anyhow!("Failed to get HEAD commit"))
  // }
}


// /// Returns (lines_added, lines_deleted) between the given commit and the current working tree.
export function get_diff_stats(&self: any, from_commit: string): any {
  // // git diff --numstat <commit>
  // let output = Command::new("git")
  // .current_dir(&self.root)
  // .args(&["diff", "--numstat", from_commit])
  // .output()?;
  // 
  // if !output.status.success() {
  // return Err(anyhow::anyhow!("Failed to run git diff"));
  // }
  // 
  // let output_str = String::from_utf8_lossy(&output.stdout);
  // let mut added = 0;
  // let mut deleted = 0;
  // 
  // for line in output_str.lines() {
  // let parts: Vec<&str> = line.split_whitespace().collect();
  // if parts.len() >= 2 {
  // if let Ok(a) = parts[0].parse::<usize>() {
  // added += a;
  // }
  // if let Ok(d) = parts[1].parse::<usize>() {
  // deleted += d;
  // }
  // }
  // }
  // 
  // Ok((added, deleted))
}


// /// Fetches the last N commits made by a specific author, including their diff patches.
// /// This is used for simulating teammate code reviews and learning their style.
export function get_commits_by_author(&self: any, author_name: string, limit: number): any {
  // let author_arg = format!("--author={}", author_name);
  // let output = Command::new("git")
  // .current_dir(&self.root)
  // .args(&["log", &author_arg, "-p", &format!("-n{}", limit)])
  // .output()?;
  // 
  // if output.status.success() {
  // Ok(String::from_utf8_lossy(&output.stdout).into_owned())
  // } else {
  // Err(anyhow::anyhow!("Failed to get commits for author {}", author_name))
  // }
}


// pub fn recent_commits_for_path(

// &self,
// rel_path: &Path,
// limit: usize,
// ) -> Result<Vec<CommitSummary>> {
// let mut revwalk = self.repo.revwalk()?;
// revwalk.push_head()?;
// let mut commits = Vec::new();

// for oid_result in revwalk {
// let oid = match oid_result {
// Ok(oid) => oid,
// Err(_) => continue,
// };
// let commit = match self.repo.find_commit(oid) {
// Ok(c) => c,
// Err(_) => continue,
// };
// if self.commit_touches_path(&commit, rel_path)? {
// commits.push(self.summarize_commit(&commit));
// if commits.len() >= limit {
// break;
// }
// }
// }

// Ok(commits)
// }

export function commit_touches_path(&self: any, commit: any, rel_path: any): any {
  // let tree = commit.tree()?;
  // let parent = match commit.parents().next() {
  // Some(p) => p,
  // None => return Ok(false),
  // };
  // let parent_tree = parent.tree()?;
  // 
  // let diff = self
  // .repo
  // .diff_tree_to_tree(Some(&parent_tree), Some(&tree), None)?;
  // 
  // let mut touched = false;
  // let rel_str = rel_path.to_string_lossy();
  // 
  // diff.foreach(
  // &mut |delta, _| {
  // if let Some(new_file) = delta.new_file().path() {
  // if new_file.to_string_lossy().contains(&*rel_str) {
  // touched = true;
  // }
  // }
  // if let Some(old_file) = delta.old_file().path() {
  // if old_file.to_string_lossy().contains(&*rel_str) {
  // touched = true;
  // }
  // }
  // if touched {
  // false
  // } else {
  // true
  // }
  // },
  // None,
  // None,
  // None,
  // )?;
  // 
  // Ok(touched)
}


export function summarize_commit(&self: any, commit: any): any {
  // let id = short_oid(commit.id());
  // let summary = commit
  // .summary()
  // .map(|s| s.to_string())
  // .unwrap_or_else(|| "<no summary>".to_string());
  // let author = commit
  // .author()
  // .name()
  // .map(|s| s.to_string())
  // .unwrap_or_else(|| "<unknown>".to_string());
  // let time = commit.time().seconds();
  // 
  // CommitSummary {
  // id,
  // summary,
  // author,
  // time,
  // }
}

// }

export function short_oid(oid: any): string {
  // let s = oid.to_string();
  // s.chars().take(8).collect()
}

